src/drivers/esp8266_drv.c

```diff
@@ -1128,38 +1176,163 @@ bool esp8266_drv_init(struct Serial *s, const int priority,
+                esp8266_close(cso->chan_id, NULL);
+        }
+
         cmd_set_check(CHECK_DATA);
+
+        xSemaphoreGive(cso->cb_semaphore);
 }
 
+/**
+ * Synchronous command used to open a new connection to a given destinataion
+ * @param proto The protocol to use.
+ * @param addr The destination address (either IP or DNS name).
+ * @param port The destination port to connect to.
+ * @return A valid Serial object if a connection was made, NULL otherwise.
+ */
 struct Serial* esp8266_drv_connect(const enum protocol proto,
-                                   const char* dst_ip,
-                                   const unsigned int dst_port)
+                                   const char* addr,
+                                   const unsigned int port)
 {
-        const int chan_id = get_next_free_channel_num();
-        if (chan_id < 0) {
+        /* This is synchronous code for now */
+        struct channel_sync_op *cso = &esp8266_state.comm.connect_op;
+        xSemaphoreTake(cso->op_semaphore, portMAX_DELAY);
+
+        struct Serial* serial = NULL;
+        cso->chan_id = get_next_free_channel_num();
+        if (cso->chan_id < 0) {
                 pr_warning(LOG_PFX "Failed to acquire a free channel\r\n");
-                return NULL;
+                goto done;
         }
 
-        struct channel *ch = get_channel_for_use(chan_id);
+        struct channel *ch = get_channel_for_use(cso->chan_id);
         if (NULL == ch) {
                 pr_warning(LOG_PFX "Can't allocate resources for channel\r\n");
-                return NULL;
+                goto done;
         }
 
-        ch->in_use = true;
-        if (!esp8266_connect(chan_id, proto, dst_ip, dst_port,
-                             _connect_cb)) {
+        bool connect_result = false;
+        switch(proto) {
+        case PROTOCOL_TCP:
+                connect_result = esp8266_connect_tcp(cso->chan_id, addr,
+                                                     port, -1, connect_cb);
+                break;
+        case PROTOCOL_UDP:
+                ;
+                int src_port = 0;
+                enum esp8266_udp_mode udp_mode = ESP8266_UDP_MODE_NONE;
+                if (0 == strcmp(addr, IPV4_BROADCAST_ADDRESS_STR)) {
+                        src_port = port;
+                        udp_mode = ESP8266_UDP_MODE_PEER_CHANGE_WHENEVER;
+                }
+                connect_result = esp8266_connect_udp(cso->chan_id, addr,
+                                                     port, src_port,
+                                                     udp_mode, connect_cb);
+                break;
+        }
+        if (!connect_result) {
                 pr_warning(LOG_PFX "Failed to issue connect command\r\n");
-                return NULL;
+                goto done;
+        }
+
+        /* Now wait for our callback to come */
+        if (!xSemaphoreTake(cso->cb_semaphore, portMAX_DELAY)) {
```

Comment:
Yep.  The AT state machine will never fail to generate a callback (it has a timeout mechanism) so this is sane.